src/dos/dos_code_page.cpp

```diff
@@ -0,0 +1,2223 @@
+// MS-DOS 6.22) contain obviously dummy size or offset values in some places;
+// ignore them and don't spam the log output with warnings.
+static const std::set<uint16_t> DummySizeValues   = {0, 1, UINT16_MAX};
+static const std::set<uint32_t> DummyOffsetValues = {0, 1, UINT32_MAX};
+
+// ***************************************************************************
+// Storage data types
+// ***************************************************************************
+
+// Number of characters we need; the real font in the CPI file can contain a
+// different number of characters.
+constexpr uint16_t NumCharactersInFont = 256;
+
+struct ScreenFont {
+	static constexpr uint32_t FullSize_8x16 = NumCharactersInFont * 16;
+	static constexpr uint32_t FullSize_8x14 = NumCharactersInFont * 14;
+	static constexpr uint32_t FullSize_8x8  = NumCharactersInFont * 8;
+
+	std::vector<uint8_t> font_8x16 = {};
+	std::vector<uint8_t> font_8x14 = {};
+	std::vector<uint8_t> font_8x8  = {};
+};
+
+using ScreenFonts = std::unordered_map<uint16_t, ScreenFont>;
+
+// ***************************************************************************
+// Persistend data storage
+// ***************************************************************************
+
+static ScreenFonts ega_font_storage = {};
+
+// ***************************************************************************
+// CPI file structure reader
+// ***************************************************************************
+
+struct FontFileHeader {
+	static constexpr auto StructName = "FontFileHeader";
+
+	uint32_t struct_offset = 0;
+
+	uint8_t id_byte       = 0;
+	std::string id_string = {};
+	uint16_t num_pointers = 0;
+	uint8_t pointer_type  = 0;
+	uint32_t fih_offset   = 0;
+};
+
+// DR-DOS format specific structure
+struct FontFileExtendedHeader {
+	static constexpr auto StructName = "FontFileExtendedHeader";
+
+	uint32_t struct_offset = 0;
+
+	uint8_t fonts_per_codepage       = 0;
+	std::vector<uint8_t> cell_size   = {};
+	std::vector<uint32_t> dfd_offset = {};
+};
+
+struct FontInfoHeader {
+	static constexpr auto StructName = "FontInfoHeader";
+
+	uint32_t struct_offset = 0;
+
+	uint16_t num_code_pages = 0;
+};
+
+struct CodePageEntryHeader {
+	static constexpr auto StructName = "CodePageEntryHeader";
+
+	uint32_t struct_offset = 0;
+
+	uint16_t cpeh_size        = 0;
+	uint32_t next_cpeh_offset = 0;
+	uint16_t device_type      = 0;
+	std::string device_name   = {};
+	uint16_t code_page        = 0;
+	uint32_t cpih_offset      = 0;
+};
+
+struct CodePageInfoHeader {
+	static constexpr auto StructName = "CodePageInfoHeader";
+
+	uint32_t struct_offset = 0;
+
+	uint16_t version   = 0;
+	uint16_t num_fonts = 0;
+	uint16_t size      = 0;
+};
+
+struct ScreenFontHeader {
+	static constexpr auto StructName = "ScreenFontHeader";
+
+	uint32_t struct_offset = 0;
+
+	uint8_t height     = 0;
+	uint8_t width      = 0;
+	uint8_t y_aspect   = 0;
+	uint8_t x_aspect   = 0;
+	uint16_t num_chars = 0;
+};
```

Comment:
For all structs: comments or links to reference documents, etc. (I'd prefer comments)

Reply:
Added comments, everywhere. Links to the format description can be found at the top of the file.